### PR TITLE
Added Enforcer class

### PR DESCRIPTION
**Pull request by jokius**

### CODE REVIEW

File: lib/casbin/core_enforcer.rb [hunk 1]
```diff
@@ -83,12 +88,23 @@ def load_model
...
+      self.fm = Casbin::Model::FunctionMap.load_function_map
+    end
+
+    # gets the current role manager.
+    def role_manager
+      rm_map['g']
+    end
+
+    # sets the current role manager.
+    def ole_manager=(rm)
```
Comment (evgeniy-r) [inline lib/casbin/core_enforcer.rb:106]: `role_manager=`?

File: lib/casbin/core_enforcer.rb [hunk 2]
```diff
@@ -62,7 +67,7 @@ def init_with_model_and_adapter(m, adapter = nil, logger: Logger.new($stdout))
 
     # creates a model.
     def self.new_model(path = '', text = '', logger: Logger.new($stdout))
-      m = Model::Model.new logger: logger
+      m = Casbin::Model::Model.new logger: logger
```
Comment (evgeniy-r) [inline lib/casbin/core_enforcer.rb:70]: Do we really need `Casbin::` everywhere?

File: lib/casbin/core_enforcer.rb [hunk 3]
```diff
@@ -140,83 +158,134 @@ def save_policy
...
+    # controls whether to save a policy rule automatically to the adapter when it is added or removed.
+    def enable_auto_save(auto_save)
+      self.auto_save = auto_save
+    end
+
+    # controls whether to rebuild the role inheritance relations when a role is added or deleted.
+    def enable_auto_build_role_links(auto_build_role_links)
+      self.auto_build_role_links = auto_build_role_links
+    end
+
```
Comment (evgeniy-r) [inline lib/casbin/core_enforcer.rb:176]: Maybe we can just use attributes (`enabled=`, `auto_save=` and `auto_build_role_links=`)?
Or this is troublesome?

File: lib/casbin/effect/allow_and_deny_effector.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+require 'casbin/effect/effector'
+
+module Casbin
+  module Effect
+    class AllowAndDenyEffector < Casbin::Effect::Effector
+      # returns a intermediate effect based on the matched effects of the enforcer
+      def intermediate_effect(effects)
+        return Casbin::Effect::Effector::DENY if effects.include?(Casbin::Effect::Effector::DENY)
+
+        Casbin::Effect::Effector::INDETERMINATE
```
Comment (evgeniy-r) [inline lib/casbin/effect/allow_and_deny_effector.rb:12]: Can we use just `INDETERMINATE`? It seems that this constant is declared in the parent class.

File: lib/casbin/enforcer.rb [hunk 5]
```diff
@@ -12,95 +15,86 @@ module Casbin
...
-      puts "call get_users_for_role with name = #{name}"
+      model.model['g']['g'].rm.get_users(name)
     end
 
     # determines whether a user has a role.
     def has_role_for_user(name, role)
-      # stub method
-      puts "call has_role_for_user with name = #{name} role = #{role}"
+      roles = get_roles_for_user(name)
+      roles.any? { |r| r == role }
```
Comment (evgeniy-r) [inline lib/casbin/enforcer.rb:29]: Can we use `roles.include?(r)`

File: lib/casbin/enforcer.rb [hunk 6]
```diff
@@ -110,15 +104,20 @@ def has_permission_for_user(user, *permission)
...
-
-    # gets the roles that a user has inside a domain.
-    def get_roles_for_user_in_domain(name, domain)
-      # stub method
-      puts "call get_roles_for_user_in_domain with name = #{name} domain = #{domain}"
+    def get_implicit_roles_for_user(name, domain = nil)
+      res = []
+      queue = Array.new([name])
+      while queue.size.positive?
+        name = queue.delete_at(0)
```
Comment (evgeniy-r) [inline lib/casbin/enforcer.rb:111]: Should we add some guard from the endless cycle?

File: lib/casbin/enforcer.rb [hunk 7]
```diff
@@ -110,15 +104,20 @@ def has_permission_for_user(user, *permission)
...
-      puts "call get_implicit_roles_for_user with name = #{name} domain = #{domain}"
-    end
-
-    # gets the roles that a user has inside a domain.
-    def get_roles_for_user_in_domain(name, domain)
-      # stub method
-      puts "call get_roles_for_user_in_domain with name = #{name} domain = #{domain}"
+    def get_implicit_roles_for_user(name, domain = nil)
+      res = []
+      queue = Array.new([name])
```
Comment (evgeniy-r) [inline lib/casbin/enforcer.rb:109]: `queue = [name]`?

File: spec/casbin/model/model_spec.rb [hunk 8]
```diff
@@ -1,339 +0,0 @@
-# frozen_string_literal: true
```
Comment (evgeniy-r) [inline spec/casbin/model/model_spec.rb:1]: Why do we delete this tests?

File: spec/casbin/persist/adapters/file_adapter_spec.rb [hunk 9]
```diff
@@ -1,159 +0,0 @@
-# frozen_string_literal: true
```
Comment (evgeniy-r) [inline spec/casbin/persist/adapters/file_adapter_spec.rb:1]: Why do we delete this tests?

File: spec/casbin/core_enforcer_spec.rb [hunk 10]
```diff
@@ -38,6 +38,8 @@
       end
```
Comment (evgeniy-r) [inline spec/casbin/core_enforcer_spec.rb:39]: Should we add new examples? Or activate the ignored ones?

File: lib/casbin/core_enforcer.rb (hunk 1 above)
Comment (jokius) [reply lib/casbin/core_enforcer.rb:106]: yes. Ty

File: lib/casbin/core_enforcer.rb [hunk 11]
```diff
@@ -30,12 +30,12 @@ def initialize(model = nil, adapter = nil, logger: Logger.new($stdout))
       end
     end
 
-    attr_accessor :adapter, :auto_build_role_links, :auto_save, :effector, :enabled, :role_manager, :watcher
+    attr_accessor :adapter, :auto_build_role_links, :auto_save, :eft, :enabled, :watcher, :rm_map, :effector
```
[the remark this answers is not in the document]
Comment (jokius) [reply lib/casbin/core_enforcer.rb:33]: renamed

File: lib/casbin/enforcer.rb (hunk 6 above)
Comment (jokius) [reply lib/casbin/enforcer.rb:111]: we cannot know the nesting depth

File: spec/casbin/model/model_spec.rb (hunk 8 above)
Comment (jokius) [reply spec/casbin/model/model_spec.rb:1]: overlapped by tests from spec/casbin/enforcer_spec.rb
And doesn't work with stubs

File: spec/casbin/core_enforcer_spec.rb (hunk 10 above)
Comment (jokius) [reply spec/casbin/core_enforcer_spec.rb:39]: More examples and tests in spec/casbin/enforcer_spec.rb

File: spec/casbin/persist/adapters/file_adapter_spec.rb (hunk 9 above)
Comment (jokius) [reply spec/casbin/persist/adapters/file_adapter_spec.rb:1]: overlapped by tests from spec/casbin/enforcer_spec.rb
And doesn't work with stubs

File: lib/casbin/core_enforcer.rb (hunk 3 above)
Comment (jokius) [reply lib/casbin/core_enforcer.rb:176]: this is used outside of the gem, I think we'll leave it that way for now

File: spec/casbin/persist/adapters/file_adapter_spec.rb (hunk 9 above)
Comment (evgeniy-r) [reply spec/casbin/persist/adapters/file_adapter_spec.rb:1]: Is saving of a policy to the file tested?

**Pull request merged**